Target label

antonrohr opened the pull request:
This makes the target label the same thing in the ProfilingTargetDialog and the Orbitmainwindow (and adds tests)

beckerhe commented on src/OrbitQt/CMakeLists.txt:
```diff
@@ -112,9 +114,9 @@ target_link_libraries(
           qtpropertybrowser::qtpropertybrowser)
 
-if(WITH_CRASH_HANDLING)
-  target_link_libraries(OrbitQt PRIVATE CrashHandler)
-  target_compile_definitions(OrbitQt PRIVATE ORBIT_CRASH_HANDLING)
+if(WITH_CRASH_HANDLING)
+  target_link_libraries(OrbitQt PRIVATE CrashHandler)
+  target_compile_definitions(OrbitQt PRIVATE ORBIT_CRASH_HANDLING)
```
What has changed here?

antonrohr replied:
Took me a while to figure it out. I think vscode or clang format fixed line endings here (CRLF -> LF). 
![Screenshot 2021-03-11 at 10 49 37](https://user-images.githubusercontent.com/7501835/110768382-7adcfa00-8257-11eb-89e1-f4dcc3619efe.png)

beckerhe replied:
So nothing changed and it can be removed from this commit?

antonrohr replied:
Well. The file contained mixed line endings before this commit. Which is a bad thing in the first place. Now its just LF line endings and thats a good thing. I could add a dedicated "Fix line endings" commit, but for the sake of simplicity, I would leave it as is. wdyt?

beckerhe replied:
I personally value clean commits a lot higher than consistent line endings, but I also don't feel strongly at all about this. So if you prefer to leave it, that's fine with me!

beckerhe commented on src/OrbitQt/TargetLabel.h:
```diff
@@ -0,0 +1,35 @@
+
+namespace orbit_qt {
+
+class TargetLabel : public QLabel {
```
I think you shouldn't inherit from QLabel here. To me it looks like the label should be a private member instead.

You're trying to enforce invariants by setting the label's properties in the setters of `TargetLabel`. But if you inherit from QLabel, the user of this class will be able to mess with this.

I see two solutions:
- Either make it a private member, so you have full control over the invariance and the user can rely on that.
- Or make it obvious that it's the user's responsibility, by making the setters free functions which take a label as an argument. `SetFileTarget(QLabel* target_label, const std::filesystem::path& path);`

Obviously the first is closer to the current solution.

reichlfl replied:
I second this, because it'll also make it easier to extend it by e.g. adding a "Connected" icon as we discussed.

antonrohr replied:
I implemented the first solution (in a way). I mark this as resolved

beckerhe commented on src/OrbitQt/TargetLabelTest.cpp:
```diff
@@ -0,0 +1,115 @@
+TEST(TargetLabel, Constructor) {
+  TargetLabel label{};
+  EXPECT_TRUE(label.text().isEmpty());
+  EXPECT_TRUE(label.styleSheet().isEmpty());
```
I'm not saying you should remove it, only to think about it if this couldn't be checked differently.

It looks a lot like an implementation detail to me. In the end for the user of the class it should be of no concern
how the font coloring is done. You could also use QPalette, or overwrite the paintEvent method, or even role your own label implementation. The only important thing in the end is the color of the text, not the stylesheet.

And by putting that check into the tests you even guarantee to the user that this is officially observable behaviour. That means someone could start checking the stylesheet property to figure out in which mode the target label currently is. Still unlikely for this use-case, but I hope you get the idea.

antonrohr replied:
As we talked offline, I changed the testing of this. It now uses QPalette instead of spreadsheet and I only test for change in color.

reichlfl commented on src/OrbitQt/TargetLabelTest.cpp:
```diff
@@ -0,0 +1,115 @@
+  TargetLabel label;
+
+  label.SetProcessAlive();
+  EXPECT_TRUE(label.styleSheet().contains(TargetLabel::kGreenColor));
+  EXPECT_TRUE(label.toolTip().isEmpty());
```
Nit: Since you test this twice (here and in `SetProcessAlive`), it would be easier to maintain and update if you'd move it to a function. Same for the three tests below.

antonrohr replied:
outdated

beckerhe commented on src/OrbitQt/TargetLabel.cpp:
```diff
@@ -0,0 +1,50 @@
+
+void TargetLabel::SetProcessDead() {
+  setStyleSheet(kStyleSheetFormat.arg(kOrangeColor));
+  setToolTip("The process ended on the instance");
```
Nit: It's a sentence, I would make a fullstop at the end.

antonrohr replied:
done

reichlfl commented on src/OrbitQt/TargetLabel.h:
```diff
@@ -0,0 +1,35 @@
+class TargetLabel : public QLabel {
+ public:
+  explicit TargetLabel(QWidget* parent = nullptr);
+  void SetFileTarget(const std::filesystem::path& path);
+  void SetStadiaTarget(const ProcessData& process, const orbit_ggp::Instance& instance);
+  void SetLocalTarget(const ProcessData& process);
```
I'm not super happy with the API here - there's 3 methods that look like regular setters, but their usage is mutually exclusive.
Maybe just rename them to make this clear, e.g. `ChangeToXXXTarget`.

antonrohr replied:
I changed it, ptal

reichlfl commented on src/OrbitQt/TargetLabel.h:
```diff
@@ -0,0 +1,35 @@
+  void SetFileTarget(const std::filesystem::path& path);
+  void SetStadiaTarget(const ProcessData& process, const orbit_ggp::Instance& instance);
+  void SetLocalTarget(const ProcessData& process);
+  void SetProcessDead();
+  void SetProcessAlive();
+  void SetConnectionDead(const QString& error_message);
```
Same as above. Maybe `SetStatus` with an enum parameter, and an optional tooltip parameter?

antonrohr replied:
I changed it, ptal

reichlfl commented on src/OrbitQt/TargetLabel.h:
```diff
@@ -0,0 +1,35 @@
+  void SetProcessAlive();
+  void SetConnectionDead(const QString& error_message);
+
+  static const QString kDefaultColor;
+  static const QString kGreenColor;
+  static const QString kOrangeColor;
+  static const QString kRedColor;
```
Nit: These should be private, and I assume they are public because of the unit test? I'd avoid making something public just to be testable - it's good practice to test through the public API and ignore internals, so this indicates you're testing internals.

However - since this whole widget has purely visual purposes, this level of detail may be fine here... I just wanted to point this out, but I'm undecided.

antonrohr replied:
I made them an implementation detail and dont test for correct color anymore, only for change in color. ptal

reichlfl reviewed (commented):
This is nice, the encapsulation makes it a lot cleaner. PTAL at my comments regarding the API, I'd suggest to at least rename the methods to sound less like setters.

beckerhe reviewed (approved):
Otherwise LGTM

antonrohr commented:
I changed quite a bit in this PR and I needed the cmake change that was submitted yesterday, hence I rebased and force pushed. @reichlfl @beckerhe please take a look again. 

This now also adds cpu utilization and icons and therefore does everything tracked in http://b/177305224

beckerhe commented on src/OrbitQt/ProfilingTargetDialog.cpp:
```diff
@@ -265,13 +264,14 @@ void ProfilingTargetDialog::SetupStadiaStates() {
+    ui_->targetLabel->ChangeToStadiaTarget(*process_,
+                                           ui_->stadiaWidget->GetSelectedInstance().value());
   });
+  QObject::connect(&state_stadia_process_selected_, &QState::exited, this,
+                   [this] { ui_->targetLabel->Clear(); });
```
why not
QObject::connect(&state_stadia_process_selected_, &QState::exited, ui_->targetLabel, &TargetLabel::Clear);?

antonrohr replied:
done

beckerhe commented on src/OrbitQt/ProfilingTargetDialog.cpp:
```diff
@@ -458,32 +459,45 @@ void ProfilingTargetDialog::OnProcessListUpdate(
+    if (ui_->processesTableView->selectionModel()->hasSelection()) {
+      const auto it =
+          std::find_if(process_list.begin(), process_list.end(), [&](const auto& process) {
+            return process.pid() == ui_->processesTableView->selectionModel()
+                                        ->currentIndex()
+                                        .data(Qt::UserRole)
+                                        .value<const ProcessInfo*>()
+                                        ->pid();
```
The right hand side of this comparison is constant, isn't it? It's hard to understand at first that this doesn't change. I'd suggest to put this pid into a variable first, and then search for the correct process by comparing to this variable.

beckerhe replied:
One more thing: The "current index" is somehow independent from a selection, and mainly meant for keyboard navigation.

You can have a selection and a different current index, or maybe no current index at all. I think you wanna ask the selection model for the selectedRows() and then take the first one.

antonrohr replied:
done

beckerhe commented on src/OrbitQt/orbitmainwindow.cpp:
```diff
@@ -1179,9 +1173,8 @@ void OrbitMainWindow::OnStadiaConnectionError(std::error_code error) {
-      kTargetLabelDefaultStyleSheet.arg(kTargetLabelColorTargetDisconnected));
-  target_label_->setToolTip(error_message);
+  target_label_->SetConnectionDead(error_message);
+  ui->menuBar->adjustSize();
```
Nit: This feels a bit like a dependency on the implemenation of TargetLabel, because you need to know which method of TargetLabel changes the widget's size. Ideally this is decoupled and TargetLabel tells you when a call of adjustSize is needed.

This could be done by a signal in TargetLabel. Maybe there is already one in QWidget which could be used. Haven't checked though.

antonrohr replied:
I like this idea a lot. QWidget did not have a signal that was useful, so I added a `SizeChanged` signal. done

beckerhe reviewed (commented):
Very Nice, Anton. I added some comments

antonrohr commented on src/OrbitQt/TargetLabel.h:
```diff
@@ -0,0 +1,68 @@
+  QString machine_ = "";
+  std::optional<Icon> icon_;
+
+  const QPixmap process_alive_icon_;
+  const QPixmap process_ended_icon_;
+  const QPixmap connection_ended_icon_;
```
They are actually not there because of the tests, but because I "compute" them. Aka take the white icon and color it pixel by pixel (see TargetDialog.cpp anon namespace `ColorizeIcon()`). I did not want to make this computation everytime I use them, but once. I test via the enum. 

The static is a good idea. I will change it to static

beckerhe replied:
But that still doesn't explain why they need to be in the class. Why not static variables in the cpp file? You could still compute them.

antonrohr replied:
done

antonrohr commented on src/OrbitQt/ProfilingTargetDialog.cpp:
```diff
@@ -397,8 +396,10 @@ void ProfilingTargetDialog::SetupFileStates() {
+    ui_->targetLabel->ChangeToFileTarget(selected_file_path_);
   });
+  QObject::connect(&state_file_selected_, &QState::exited, this,
+                   [this] { ui_->targetLabel->Clear(); });
```
done

antonrohr commented on src/OrbitQt/ProfilingTargetDialog.cpp:
```diff
@@ -341,11 +340,12 @@ void ProfilingTargetDialog::SetupLocalStates() {
+    ui_->targetLabel->ChangeToLocalTarget(*process_);
   });
+  QObject::connect(&state_local_process_selected_, &QState::exited, this,
+                   [this] { ui_->targetLabel->Clear(); });
```
Thank you! Missed that. done

antonrohr commented on src/OrbitQt/TargetLabel.h:
```diff
@@ -0,0 +1,68 @@
+  [[nodiscard]] QColor GetColor() const;
+  [[nodiscard]] QString GetText() const;
+  [[nodiscard]] QString GetToolTip() const { return toolTip(); }
+  [[nodiscard]] std::optional<Icon> GetIcon() const { return icon_; }
```
done

antonrohr commented on src/OrbitQt/TargetLabel.cpp:
```diff
@@ -0,0 +1,171 @@
+  if (process_.isEmpty() || machine_.isEmpty()) return false;
+
+  ui_->textLabel->setText(
+      QString{"%1 (%2%) @ %3"}.arg(process_, QString::number(cpu_usage, 'f', 0), machine_));
```
Did not know about that arg overload. neat! done

antonrohr commented on src/OrbitQt/TargetLabel.cpp:
```diff
@@ -0,0 +1,171 @@
+  SetProcessCpuUsage(cpu_usage);
+}
+
+bool TargetLabel::SetProcessCpuUsage(double cpu_usage) {
```
done

antonrohr commented on src/OrbitQt/TargetLabel.h:
```diff
@@ -0,0 +1,68 @@
+class TargetLabel : public QWidget {
+  Q_OBJECT
+ public:
+  enum class Icon { kProcessAlive, kProcessEnded, kConnectionDead };
```
I renamed it to IconType. For easy reviewing its a separate fixup rename commit. ptal

beckerhe replied:
Sounds good.

antonrohr commented on src/OrbitQt/TargetLabel.h:
```diff
@@ -0,0 +1,68 @@
+ private:
+  std::unique_ptr<Ui::TargetLabel> ui_;
+  QString process_ = "";
+  QString machine_ = "";
```
done

beckerhe commented on src/OrbitQt/orbitmainwindow.cpp:
```diff
@@ -5,6 +5,8 @@
 #include "orbitmainwindow.h"
 
 #include <absl/flags/declare.h>
+#include <qmenubar.h>
+#include <qobject.h>
```
Nit

antonrohr replied:
One day I will find time to setup auto fixing iwyu on every save. This gets really annoying. Thank you for pointing it out. done (and rebased)

beckerhe commented on src/OrbitQt/TargetLabel.cpp:
```diff
@@ -4,6 +4,8 @@
 
 #include "TargetLabel.h"
 
+#include <qpixmap.h>
+
```
Nit

beckerhe reviewed (approved):
Thanks a lot for the fixes. LGTM

antonrohr commented:
@reichlfl I think we are good here code wise with Hennings review, maybe you could just look at these screenshots to make sure we are on the same page.

The blurriness in the red color is an encoding artifact of chrome remote desktop. I dont know what the encoders problem with red color is, but its definitely sharp without CRD.

![Screenshot 2021-03-18 at 10 56 00](https://user-images.githubusercontent.com/7501835/111607825-e6374680-87d8-11eb-8341-b060a040553b.png)
![Screenshot 2021-03-18 at 10 55 00](https://user-images.githubusercontent.com/7501835/111607832-e7687380-87d8-11eb-943d-064b12d7ed6e.png)
![Screenshot 2021-03-18 at 10 55 20](https://user-images.githubusercontent.com/7501835/111607835-e8010a00-87d8-11eb-89ae-7a65d0496de1.png)
![Screenshot 2021-03-18 at 10 54 54](https://user-images.githubusercontent.com/7501835/111607838-e8010a00-87d8-11eb-9f7b-3896934cb49a.png)
![Screenshot 2021-03-18 at 10 54 42](https://user-images.githubusercontent.com/7501835/111607841-e899a080-87d8-11eb-9950-f395ad4b88e1.png)

reichlfl commented:
> @reichlfl I think we are good here code wise with Hennings review, maybe you could just look at these screenshots to make sure we are on the same page.

Sorry, I didn't realize I had open discussion items here. I'll trust in your and Hennings judgment on this, no need to wait for me to do a second review.

Regarding the screenshots: Really nice, I like the inclusion of the icon. A nitpick, the red looks really aggressive to a point where it's a bit hard to read on the grey background, maybe it should be a bit of a lighter shade. But that might also be due to the compression.

Orange indicates the process connection was lost, but the instance is still alive? A different icon for this could be nice, but not urgently needed. Could also be changed at a later point. Do we have tooltips that indicate the difference between the red and orange state? It's not clear from just looking at the label / icon.

Maybe we just use a "Warning" icon when the process connection is lost? https://material.io/resources/icons/?search=problem&icon=report_problem&style=outline
In the future, I can imagine you can still take a capture as we more and more become a system profiler (e.g. memory, scheduling and GPU information is still valid), so "Warning" may be appropriate.

antonrohr commented:
Thanks for the suggestions Florian. I changed the icon and made the red a bit brighter. There are also tooltips. The process has ended one could be a bit more verbose, in case you have better idea. 

![Screenshot 2021-03-18 at 13 31 58](https://user-images.githubusercontent.com/7501835/111628081-1807d780-87f0-11eb-86a9-47282e2297b6.png)
![Screenshot 2021-03-18 at 13 41 29](https://user-images.githubusercontent.com/7501835/111628083-19390480-87f0-11eb-9cb3-ccfcf90aa396.png)
![Capture](https://user-images.githubusercontent.com/7501835/111628154-3077f200-87f0-11eb-8013-d258c4742a4a.PNG)

reichlfl commented:
> Thanks for the suggestions Florian. I changed the icon and made the red a bit brighter. There are also tooltips. The process has ended one could be a bit more verbose, in case you have better idea.
> 

More nitpicking, I'd probably make the icons a bit smaller - but this is really a minor detail at this point.
Can we already automatically reconnect to the process? Then I'd state that in the tooltip (e.g. "Restart a process with a matching name to continue profiling").

antonrohr commented:
I made the icons smaller. 

The reconnection to a restarted process is not yet implemented. The actual change is just a couple of lines, but I am somewhat afraid to break something at a different.  This will be a different PR. (I will remember to adjust the tooltip)

![Screenshot 2021-03-18 at 16 33 25](https://user-images.githubusercontent.com/7501835/111652940-c3705680-8807-11eb-9b19-245562ac31ea.png)
![Screenshot 2021-03-18 at 16 32 57](https://user-images.githubusercontent.com/7501835/111652943-c408ed00-8807-11eb-81d6-c1abf4ff8f6e.png)

reichlfl commented:
> I made the icons smaller.
> 
> The reconnection to a restarted process is not yet implemented. The actual change is just a couple of lines, but I am somewhat afraid to break something at a different. This will be a different PR. (I will remember to adjust the tooltip)

Makes sense, thanks!
